Update meow dependency

ericcornelissen opened the pull request.

LitoMore commented on package.json:
```diff
@@ -2,6 +2,7 @@
   "version": "1.0.5",
   "description": "Linter for SVGs",
+  "type": "module",
   "main": "src/svglint.js",
```
```suggestion
  "exports": "./src/svglint.js"
```

LitoMore replied:
See: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c#how-can-i-move-my-commonjs-project-to-esm

LitoMore commented:
Readme also needs to be updated.

ericcornelissen commented:
The current status of this Pull Request:

- Refactored the source code from CommonJS to ESModules.
- Updated Meow to a recent version.
- The CLI is unchanged except for the fact that it now has a higher minimum NodeJS version requirement due to the source code being ESM.
- [The JavaScript API](https://github.com/birjolaxew/svglint#usage) has changed from sync to `async` due to [this change](https://github.com/birjolaxew/svglint/pull/41/files#diff-6d4280a764d0c826513176f6efe43533da3d5a6b63245a4ce7f84c85176f615eR33). In short, dynamically importing ESM appears to require `async`. Hence, to support ESM plugins it appears `async` is required.
- Since the previous points already require a major version bump, I did not yet bother to look into providing a CommonJS version of either the CLI or JS API. Especially considering the fact that a transpiled CommonJS version of SVGLint would not be able to use ESM plugins, so from that viewpoint I'm also not sure if we even want to provide that option.
- It addresses #34, #36, and likely #44.

This means that a major version bump would be required to release this - for the NodeJS version requirement **and** the lack of CJS support **and** the JS API change.

@birjolaxew do you have any opinion on this? At the moment my opinion is leaning towards this solution - a mostly ESM-only solution with backwards compatibility support for CJS - because 1) it's simpler to implement and 2) it avoids any confusion due to someone trying to import an ESM config/plugin from the transpiled version of SVGLint.

Also, a correction to [what I said before](https://github.com/birjolaxew/svglint/issues/34#issuecomment-872571645): not all v12 release of NodeJS support ESM, I _believe_ the first version with proper support is v12.20.0 (the latest v12, as of this comment, is v12.22.7).

LitoMore commented:
The first version with proper support of ESM is v12.17.0. But `meow` and most of [Sindre](https://github.com/sindresorhus)'s packages are using [`node:` protocol](https://nodejs.org/api/esm.html#node-imports) for imports, which is requiring v12.20.0.

Here is the discussion: https://github.com/sindresorhus/meta/discussions/15#discussioncomment-688195.

LitoMore commented on .github/workflows/test.yml:
```diff
@@ -13,7 +16,7 @@ jobs:
       - name: Node.js setup
         uses: actions/setup-node@v1
```
Feel free to use `actions/setup-node@v2`.

```suggestion
        uses: actions/setup-node@v2
```

ericcornelissen replied:
I don't mind upgrading but I'm considering that to be out of scope for this Pull Request. E.g., when changing this I would also consider changing `runs-on: ubuntu-18.04` to something more recent. But I'd like to avoid that discussion in this thread. Feel free to open a Pull Request for this though :slightly_smiling_face: 

To clarify, I consider the other change int this file in scope because I want some certainty that everything actually works on these version of Node.

LitoMore replied:
np. :+1:

ericcornelissen replied:
I'm planning to address anything CI related in https://github.com/birjolaxew/svglint/pull/48, feel free to chime in with more suggestions if you want @LitoMore :slightly_smiling_face:

LitoMore commented on package.json:
```diff
@@ -39,7 +39,7 @@
   },
   "engines": {
-    "node": "^12.20.0 || ^14.13.1 || >=16.0.0"
+    "node": "^12.17.0 || ^14.13.1 || >=16.0.0"
```
It should be `^12.20.0 || ^14.13.1 || >=16.0.0`. Because we are using Meow, it requires `^12.20.0 || ^14.13.1 || >=16.0.0`.

ericcornelissen replied:
Ah damn, of course :sweat_smile: thanks for catching that mistake!

LitoMore commented on .github/workflows/test.yml:
```diff
@@ -5,6 +5,12 @@ jobs:
     runs-on: ubuntu-18.04
+    strategy:
+      matrix:
+        node-version:
+          - 12.17.0
+          - 14.13.1
+          - 16.0.0
```
It's unnecessary to target versions. It will only test the specific versions that you provide. Then you will miss 12.18.0, 14.14.0, 16.1.0, 17.0.0 and others.

We can target versions like this:

```suggestion
        node-version:
          - ^12.17
          - ^14.13.1
          - ^16
          - ^17
```

LitoMore replied:
BTW, there is no difference between `^12.17, ^14.13.1, ^16, ^17` and `12, 14, 16, 17`. `setup-node` will pick the latest version from the scope that you provide. Both are OK.

ericcornelissen replied:
Not sure if I'm 100% in agreement with you here. I want to test on specifically the lowest versions `package.json` says we're supporting because if I don't I can't be sure those actually work. E.g. I may start using a language feature introduced in `14.14.0`, which would break for users complying with the `engines` field, but I wouldn't know it until after the changes was released.

You could try to argue the same for higher versions, but we have to assume Node is doing it's job correctly and any version `^14.13.1` works if `14.13.1` works.

That being said, I completely agree we should include a v17 here (latest seems fine to me). As for v16, on the one hand that "feels" fine to me to test both `16.0.0` and `^16`. On the other hand, to be consistent we should then also test `12.17.0` and `^12` as well as `14.13.1` and `^14`.

Given all that, I'd instead suggest either:

```suggestion
        node-version:
          # Lowest versions we say we support (see package.json)
          - 12.20.0
          - 14.13.1
          - 16.0.0
          # Highest version we're interested in
          - ^17
```

Or:

```suggestion
        node-version:
          # Lowest versions we say we support (see package.json)
          - 12.20.0
          - 14.13.1
          - 16.0.0
          # Highest versions we're interested in
          - ^12
          - ^14
          - ^16
          - ^17
```

(with 8db814737b06d8d9095fdef5bf1850b50eada4a6, the v12 test target should be updated to `12.20.0`)

LitoMore replied:
`12.20.0, 14.13.1, 16.0.0, 12, 14, 16, 17` looks good. `^` is unnecessary but it's up to you.

birjolaxew commented:
@ericcornelissen I unfortunately don't know enough about CJS/ESM support to really weigh in on this (nor the time to read up on it). Although breaking changes (especially when it comes to dropping support for Node versions) are frustrating, I trust that the pros outweighs the cons if that's the general opinion. I'd be happy with this solution.

sachinraja commented on src/cli/config.js:
```diff
@@ -31,6 +31,6 @@ function getConfigurationFile(filename=".svglintrc.js", folder=process.cwd()) {
 }
 
-module.exports = {
+export default {
     getConfigurationFile,
 };
```
I think this should be a named export

sachinraja commented on src/lib/parse.js:
```diff
@@ -4,32 +4,34 @@
+}
+
+export default {
     /**
      * Clones an AST by re-parsing it's source
      * @param {AST} ast The AST to clone
      * @returns {AST} The cloned AST
      */
     clone(ast) {
         // @ts-ignore
-        return module.exports.parseSource(ast.source);
+        return parseSource(ast.source);
     },
```
these could be named exports too

sachinraja commented:
Not sure if the rules should be named exports, but there are a few other places too.

ericcornelissen commented:
> I unfortunately don't know enough about CJS/ESM support to really weigh in on this (nor the time to read up on it). Although breaking changes (especially when it comes to dropping support for Node versions) are frustrating, I trust that the pros outweighs the cons if that's the general opinion. I'd be happy with this solution.

No problem and thanks for the reply nonetheless, I was mainly interested in if you're _okay_ with the major version change for this reason. I'll be doing some further testing to decrease the chances of any breaking changes affecting users that are using a compatible NodeJS version

> Not sure if the rules should be named exports, but there are a few other places too.

Thanks for the feedback @sachinraja - I agree some of these are better as named exports, however my aim with this PR is to be a straightforward translation from CJS to ESM (+necessary changes to keep things working AND updating `meow`). Since CJS `module.exports` are equivalent to ESM `default` exports I stuck with that. Given there's already 4 things in the scope of this Pull Request I intend to exclude the refactoring w.r.t. export/import styles.

sachinraja commented:
> Given there's already 4 things in the scope of this Pull Request I intend to exclude the refactoring w.r.t. export/import styles.

This looks good to me then :+1:

ericcornelissen commented:
I ended up trying to provide a CJS version of the API and I think I succeeded with f3e91a1069190fd70147707068751924f209b3f0. As a result this would now only be a breaking changes w.r.t.:

- The minimum NodeJS version required to use the CLI.
- The API, which went from `sync` to `async`.

On the positive side, this would remove the need to migrate for API users. On the negative side, going full ESM in the future will require an additional major version bump (whether that's because we _want_ to or because a not-CLI-only dependency goes ESM only).

I'm personally of the opinion that we should go with the CJS support for now so that migrating is as easy as possible now and that the additional major version bump in the future isn't a huge deal. Does anyone in this thread have a different opinion?

sachinraja commented:
Just cloned locally and ran `node bin/cli.js icon.svg` and it worked for node 16, latest node 14 (14.18.2), and latest node 12. But I did get this error with node 14.0.0 < 14.8.0 < 14.13.1:
```
internal/process/esm_loader.js:74
    internalBinding('errors').triggerUncaughtException(
                              ^

Error [ERR_UNSUPPORTED_ESM_URL_SCHEME]: Only file and data URLs are supported by the default ESM loader
    at Loader.defaultResolve [as _resolve] (internal/modules/esm/resolve.js:781:11)
    at Loader.resolve (internal/modules/esm/loader.js:85:40)
    at Loader.getModuleJob (internal/modules/esm/loader.js:229:28)
    at ModuleWrap.<anonymous> (internal/modules/esm/module_job.js:51:40)
    at link (internal/modules/esm/module_job.js:50:36) {
  code: 'ERR_UNSUPPORTED_ESM_URL_SCHEME'
}
```

LitoMore commented:
@sachinraja You should not use `node bin/cli.js` to run the bin file. Please use these commands below to test:

```shell
npm link
svglint icon.svg

# Once you want to unlink
npm un -g svglint
```

LitoMore commented:
@ericcornelissen You could refer to [this test](https://github.com/sindresorhus/meow/blob/main/test/test.js#L583) if you want to test ESM behaviors.

LitoMore reviewed (approved):
Looks good to me. It would be better if the test can be improved.

sachinraja commented:
> Or just run ./bin/cli.js icons.svg directly.
It shouldn't matter whether I run it with node or directly, but I get the same error with that anyway. I also get it when I link the package.

LitoMore commented:
> It shouldn't matter whether I run it with node or directly.

Oh yes, you are right. I think you are using an unsupported Node.js version. Please refer to the `engines` field in `package.json`. Users with an unsupported version will get a warning or error when they try to install our package.

FYI: https://docs.npmjs.com/cli/v8/configuring-npm/package-json#engines

sachinraja commented:
> Please refer to the engines field in package.json, we have already specified them in there. Users with an unsupported version will get a warning or error when they try to install our package.

Ah you're right, didn't notice that change. `npm` doesn't seem to show the error when I just run `npm install` in the repo, but it does work when I try to install it into another project.

ericcornelissen commented:
Thanks so much for all the help @LitoMore and @sachinraja! 😄 

> Looks good to me. It would be better if the test can be improved.

I agree they can be improved in the way you suggested, but, once again, I'd like to keep that for a later PR. For now I just want to be sure that running the CLI under normal circumstances on all major OSes and with all supported NodeJS versions _works_ - i.e. doesn't error - I don't care so much about the actual output yet. It's currently just a smoke test.

Also, just to be explicit about it, running with the `--help` or `--version` flag would not be a sufficient test as the CLI exists early with those flags. So, while it would be good to test that the CLI works with those flags, it's more important to run the CLI with (and without) a configuration file and SVGs to lint.

> Ah you're right, didn't notice that change. npm doesn't seem to show the error when I just run npm install in the repo, but it does work when I try to install it into another project.

I added `.npmrc` in 047242bed22c8b16c796d3f3eb650e80967bd119 to prevent confusion like this in the future. Thanks for helping to figure out the problem @LitoMore 😃

LitoMore commented:
> Also, just to be explicit about it, running with the --help or --version flag would not be a sufficient test as the CLI exists early with those flags.

I just give a simple demo to show how to use `execa`. 😄

ericcornelissen commented:
Thanks once again @LitoMore and @sachinraja for feedback and testing :smiley:

I'll hopefully be able to release this tomorrow (i.e. December 19, 2021).
